xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/async/MapBasedLinkedBlockingQueue.java

```diff
@@ -0,0 +1,270 @@
+     */
+    public MapBasedLinkedBlockingQueue(ConcurrentMap<Long, T> internalMap)
+    {
+        super();
+        this.internalMap = internalMap;
+        this.highestKey = new AtomicLong(computeHighestKey());
+
+        // if the map contains some element, we want to populate the actual queue with those info.
+        this.populateUnderlyingQueue();
+    }
+
+    /**
+     * Compute from the internal map what's the highest key number.
+     *
+     * @return A {@code Long} corresponding to the highest key value of the Map or 0 if none is found.
+     */
+    private Long computeHighestKey()
+    {
+        return this.internalMap.keySet()
+            .stream()
+            // Sort in descending order to get the highest element first
+            .sorted((a, b) ->
+            {
+                if (a.equals(b)) {
+                    return 0;
+                } else if (a < b) {
+                    return 1;
+                } else {
+                    return -1;
+                }
+            })
+            .limit(1).findFirst().orElse(0L);
+    }
+
+    /**
+     * Populate the internal queue representation with the information coming from the Map.
+     * The idea is to be able to reuse entirely the original {@link LinkedBlockingQueue} method without really needing
+     * the map.
+     */
+    private void populateUnderlyingQueue()
+    {
+        this.internalMap.entrySet()
+            .stream()
+            .sorted((a, b) ->
+            {
+                Long keyA = a.getKey();
+                Long keyB = b.getKey();
+
+                if (keyA.equals(keyB)) {
+                    return 0;
+                } else if (keyA < keyB) {
+                    return -1;
+                } else {
+                    return 1;
+                }
+            })
+            .forEach((entry) ->
+            {
+                try {
+                    // use super to avoid putting back in the internal map the info already present
+                    super.put(entry.getValue());
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+    }
+
+    /**
+     * Put an element in the internal map and with the incremented highest key.
+     *
+     * @param t an element to add in the internal map.
+     */
+    private void internalPut(T t)
+    {
+        this.internalMap.put(this.highestKey.incrementAndGet(), t);
+    }
+
+    /**
+     * Remove an element from the internal map and decrement the highest key only if there's no more element
+     * associated to this key.
+     *
+     * @param t the element to remove.
+     */
+    private void internalRemove(Object t)
+    {
+        if (t != null) {
+            this.internalMap.remove(t);
+
+            // decrement the key if we removed the highest key element.
+            if (!this.internalMap.containsKey(this.highestKey.get())) {
```

Comment:
I don't understand this logic. What's removed is the first element of the queue so the element associated with the lowest key and not the highest.

Reply:
It's an internal remove method to be called whenever an element is removed from the queue, so it might be the head or the tail of the queue. Since I don't want to make prediction, I'm just checking that the highestkey is still present or not to update it.

Reply:
My main issue here is that you don't know the key so you can't remove the object from the map, `this.internalMap.remove(t);` makes no sense since ti is a value and not a key.